datafusion/physical-plan/src/lib.rs

```diff
@@ -446,6 +450,123 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
     }
 }
 
+/// Describes the execution mode of an operator's resulting stream with respect
+/// to its size and behavior. There are three possible execution modes: `Bounded`,
+/// `Unbounded` and `PipelineBreaking`.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum ExecutionMode {
+    /// Represents the mode where generated stream is bounded, e.g. finite.
+    Bounded,
+    /// Represents the mode where generated stream is unbounded, e.g. infinite.
+    /// Even though the operator generates an unbounded stream of results, it
+    /// works with bounded memory and execution can still continue successfully.
+    Unbounded,
+    /// Represents the mode where some of the operator's input stream(s) are
+    /// unbounded; however, the operator cannot generate streaming results from
+    /// these streaming inputs. In this case, the execution mode will be pipeline
+    /// breaking, e.g. the operator requires unbounded memory to generate results.
+    PipelineBreaking,
```

Comment:
What is the usecase of `PipelineBreaking` ? I didn't see it used anywhere in the code differently than `ExecutionMode::Bounded` (though I may have missed)

I guess I am wondering what is (practical)  difference between `Bounded` and `PipelineBreaking`?

Reply:
Some physical optimizer rules such as `replace_with_order_preserving_variants` behaves differently for `PipelineBreaking` and `PipelineFriendly` modes. (Such as choosing `SortPreservingRepartition` instead of `Repartition` to resolve `PipelineBreaking` plan).